File: java/course-schedule.java
package java;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

class CourseSchedule {
    /**
     * 使用一个入度表，和邻接数组
     */
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegrees = new int[numCourses];
        List<List<Integer>> ajacency = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            ajacency.add(new ArrayList<>());
        }
        for (int i = 0; i < prerequisites.length; i++) {
            indegrees[prerequisites[i][0]]++;
            ajacency.get(prerequisites[i][1]).add(prerequisites[i][0]);
        }
        Deque<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegrees[i] == 0) queue.offer(i);
        }
        while (!queue.isEmpty()) {
            int pre = queue.pollFirst();
            numCourses -= 1;
            for (int cur : ajacency.get(pre)) {
                indegrees[cur]--;
                if (indegrees[cur] == 0) queue.offer(cur);
            }
        }
        return numCourses == 0;
    }
}

File: python/course-schedule.py
from collections import deque
from typing import List

def canFinish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    '''
    1. 入度表
    2. 有向无环图
    这里我用入度表来解
    '''
    indegrees = [0 for _ in range(numCourses)]
    ajacency = [[] for _ in range(numCourses)]
    for course, pre in prerequisites:
        indegrees[course] += 1
        ajacency[pre].append(course)
    queue = deque()
    # 将没有前置课程的课入队
    for i in range(numCourses):
        if not indegrees[i]: queue.append(i)
    while queue:
        pre = queue.popleft()
        numCourses -= 1
        for cur in ajacency[pre]:
            indegrees[cur] -= 1
            if not indegrees[cur]: queue.append(cur)
    return not numCourses
